File: NextPermuation.py
"""
Time Complexity : O(n) where n is number of elements in nums array
Space Complexity : O(1)
Did this code successfully run on Leetcode : Yes
Any problem you faced while coding this : No
"""
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        if len(nums) == 0:
            return 0
        n = len(nums)
        temp = -1
        # Step 1: Loacate the smaller number
        i = n - 2
        while i >= 0:
            if nums[i] < nums[i + 1]:
                temp = i
                break
            i -= 1
        # Step 2: Swap the number larger and reverse the later half
        j = n - 1
        if temp != -1:
            while j >= 0:
                if nums[j] > nums[temp]:
                    nums[j], nums[temp] = nums[temp], nums[j]
                    break
                j -= 1
        self.reverse(nums, temp + 1, n - 1)
    # Reverse function
    def reverse(self, nums, l, r):
        while l < r:
            nums[l], nums[r] = nums[r], nums[l]
            l += 1
            r -= 1
            
        

File: NextPermuation.java
// Time Complexity :O(N^2)
// Space Complexity :O(1)
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this :No

class Solution {
    public void nextPermutation(int[] nums) {
        if(nums == null || nums.length==0) return;
        int  n = nums.length;
        int i = n-2;
        while(i>=-1){
            if(i==-1){
                reverse(0,n-1,nums);
                break;
            }
            if(nums[i+1]>nums[i]){
                swap(i,nums);
                break;
            }
            i--;
        }
        if(i!=-1)reverse(i+1,n-1,nums);
        
        
        
    }
    
    private void swap(int index , int[] nums){
        for(int i = nums.length-1;i>index;i--){
            if(nums[i]>nums[index]){
                int temp = nums[i];
                nums[i] = nums[index];
                nums[index] = temp;
                break;
            }
        }
    }
    private void reverse(int start,int end,int[] nums){
        while(start<end){
            int temp  = nums[start];
            nums[start] = nums[end];
            nums[end] = temp;
            end--;
            start++;
        }
    }
}